nameresolution/consul/consul.go

```diff
@@ -66,28 +65,142 @@ type healthInterface interface {
 }
 
 type resolver struct {
-	config resolverConfig
-	logger logger.Logger
-	client clientInterface
+	config   resolverConfig
+	logger   logger.Logger
+	client   clientInterface
+	registry registryInterface
+}
+
+type registryInterface interface {
+	get(service string) *registryEntry
+	expire(service string) // clears slice of instances
+	remove(service string) // removes entry from registry
+	addOrUpdate(service string, services []*consul.ServiceEntry)
+}
+
+type registry struct {
+	entries map[string]*registryEntry
+	mu      sync.RWMutex
+}
+
+type registryEntry struct {
+	services []*consul.ServiceEntry
+	mu       sync.RWMutex
+}
+
+func (r *registry) get(service string) *registryEntry {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.entries[service]
+}
+
+func (e *registryEntry) next() *consul.ServiceEntry {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if len(e.services) == 0 {
+		return nil
+	}
+
+	return shuffle(e.services)[0]
```

Comment:
Instead of shuffling to pick one, can you simply pick a random position? What is the advantage of shuffling?

Reply:
This is based on a fisher-yates shuffle, it's a non-deterministic method of generating a random result. Consul use this same approach in their dns

Reply:
To shuffle, sure. In this case, it is to pick one - which a simple random number with mod operation should accomplish the same.

Reply:
I've made the change - is that what you had in mind?